HeterogeneousCore/SonicTriton/interface/TritonClient.h

```diff
@@ -0,0 +1,93 @@
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "HeterogeneousCore/SonicCore/interface/SonicClientSync.h"
+#include "HeterogeneousCore/SonicCore/interface/SonicClientPseudoAsync.h"
+#include "HeterogeneousCore/SonicCore/interface/SonicClientAsync.h"
+
+#include <vector>
+#include <string>
+#include <exception>
+
+#include "request_grpc.h"
+
+namespace ni = nvidia::inferenceserver;
+namespace nic = ni::client;
+
+template <typename Client>
+class TritonClient : public Client {
+public:
+  struct ServerSideStats {
+    uint64_t request_count_;
+    uint64_t cumul_time_ns_;
+    uint64_t queue_time_ns_;
+    uint64_t compute_time_ns_;
+  };
+
+  //constructor
+  TritonClient(const edm::ParameterSet& params);
+
+  //helper
+  bool getResults(const std::unique_ptr<nic::InferContext::Result>& result);
+
+  //accessors
+  unsigned nInput() const { return nInput_; }
+  unsigned nOutput() const { return nOutput_; }
+  unsigned batchSize() const { return batchSize_; }
+  bool verbose() const { return verbose_; }
+  void setBatchSize(unsigned bsize) { batchSize_ = bsize; }
+
+  //for fillDescriptions
+  static void fillPSetDescription(edm::ParameterSetDescription& iDesc) {
+    edm::ParameterSetDescription descClient;
+    descClient.add<unsigned>("nInput");
+    descClient.add<unsigned>("nOutput");
+    descClient.add<unsigned>("batchSize");
+    descClient.add<std::string>("address");
+    descClient.add<unsigned>("port");
+    descClient.add<unsigned>("timeout");
+    descClient.add<std::string>("modelName");
+    descClient.add<int>("modelVersion", -1);
+    descClient.add<bool>("verbose", false);
+    descClient.add<unsigned>("allowedTries", 0);
+    iDesc.add<edm::ParameterSetDescription>("Client", descClient);
+  }
+
+protected:
+  unsigned allowedTries() const override { return allowedTries_; }
+
+  void evaluate() override;
+
+  //helper for common ops
+  bool setup();
+
+  //helper to turn triton error into warning
+  bool wrap(const nic::Error& err, const std::string& msg) const;
+
+  void reportServerSideStats(const ServerSideStats& stats) const;
+  ServerSideStats summarizeServerStats(const ni::ModelStatus& start_status, const ni::ModelStatus& end_status) const;
+
+  ni::ModelStatus getServerSideStatus() const;
+
+  //members
+  std::string url_;
+  unsigned timeout_;
+  std::string modelName_;
+  int modelVersion_;
+  unsigned batchSize_;
+  unsigned nInput_;
+  unsigned nOutput_;
+  bool verbose_;
+  unsigned allowedTries_;
+  std::unique_ptr<nic::InferContext> context_;
+  std::unique_ptr<nic::ServerStatusContext> serverCtx_;
+  std::shared_ptr<nic::InferContext::Input> nicInput_;
+};
+
+typedef TritonClient<SonicClientSync<std::vector<float>>> TritonClientSync;
+typedef TritonClient<SonicClientPseudoAsync<std::vector<float>>> TritonClientPseudoAsync;
+typedef TritonClient<SonicClientAsync<std::vector<float>>> TritonClientAsync;

```

Comment:
How about
```suggestion
using TritonClientSync = TritonClient<SonicClientSync<std::vector<float>>>;
using TritonClientPseudoAsync = TritonClient<SonicClientPseudoAsync<std::vector<float>>>;
using TritonClientAsync = TritonClient<SonicClientAsync<std::vector<float>>>;
```
?

Reply:
Not opposed, but what is the particular benefit?

Reply:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rt-using

Reply:
> **Reason** Improved readability: With `using`, the new name comes first rather than being embedded somewhere in a declaration. Generality: using can be used for template aliases, whereas typedefs can’t easily be templates. Uniformity: `using` is syntactically similar to `auto`.

Reply:
IIUC the c++ standard, the `typedef` or `using` statements do not count as explicit template instantiation declarations.

So there should also be the lines
```c++
// explicit template instantiation declarations
extern template class TritonClient<SonicClientSync<std::vector<float>>>;
extern template class TritonClient<SonicClientAsync<std::vector<float>>>;
extern template class TritonClient<SonicClientPseudoAsync<std::vector<float>>>;
```
to match the definitions in the .cc file.

Reply:
I was under the impression that `extern template` was intended to *prevent* implicit instantiation, because there is an explicit instantiation elswhere: https://isocpp.org/wiki/faq/cpp11-language-templates. Since the class template introduced here is not fully implemented, there should not be any chance of implicit instantiation. But perhaps my understanding is incomplete.

Reply:
But you do have an explicit template instantiation elsewhere, at the bottom of the `HeterogeneousCore/SonicTriton/src/TritonClient.cc` file.

What am I missing ?

Reply:
My point is that *preventing* implicit instantiation in the header file is not necessary, because the class cannot be compiled using just the information in the header file. `extern template` only affects compilation time, as far as I understand.

I tried compiling a few times with and without these lines, and I don't noticed a significant difference in the compilation time.

(Also, for future reference, if these lines are added, `template <> void TritonClientAsync::evaluate();` is needed before them to avoid an "explicit specialization after instantiation" error.)

Reply:
> My point is that _preventing_ implicit instantiation in the header file is not necessary, because the class cannot be compiled using just the information in the header file.

The template class can be instantiated, and calling inlined functions works fine. Calling out-of-line functions also compiles, but leads to linker errors if the out-of-line function definition (for the template argument) is not visible to the linker.

Reply:
Is this comment in support of using `extern template class`? (I only find one other instance of it in all of CMSSW.)

Reply:
"maybe"? I don't have clear preference in either way, just wanted to point out that a class template with just function declarations can be instantiated, and everything seems to work "fine" until a declared-only function is called.

Reply:
I was curious, so I ran the compilation from scratch 100 times with and without this change on a dedicated machine. It decreases the compile time for this package by 4% on average, with a significance of 4.3 sigma. So it's definitely a micro-optimization in this case, but it does have an impact. I'll make the requested change.